### PR TITLE
Use lazy-loaded cached tree implementation

### PR DESCRIPTION
**Pull request by roysc**
This introduces a largely rewritten implementation of the SMT which operates on a cached, lazily-loaded tree structure, as opposed to the current paradigm of reading from/writing to the DB directly for each operation. This gives a significant performance [improvement](https://github.com/cosmos/cosmos-sdk/issues/11444#issuecomment-1103714044) on all operations while preserving full compatibility of the produced hashes.

This also refactors the hasher objects to allow tree paths and stored values to be hashed independently, and `Option`s to configure them. https://github.com/celestiaorg/smt/issues/40 can be satisfied by passing an identity function as the path hasher.

As noted [here](https://github.com/cosmos/cosmos-sdk/issues/11444#issuecomment-1112401379), the `DeepSubtree` code supporting state-transition fraud proofs was not compatible as-is with this implementation and so is removed; but there should be no technical block to adapting that code to this impl.; it was simply outside the scope of Cosmos-SDK ADR-40 for which this was originally built.

### CODE REVIEW

**Comment by musalbas** [comment]
Thanks @roysc! I've pulled the changes from the vulcanize repo a few weeks into the "lazy" branch here (https://github.com/celestiaorg/smt/tree/lazy), as we need to implement the things in https://github.com/celestiaorg/smt/issues/72 first before merging to master. Is our lazy branch already up to date or does this PR introduce anything new not in the lazy branch (if so can this PR be against the lazy branch rather than master)?

For context though, we've currently decided to put the SMT work on hold too, as we're currently implementing deep trees for IAVL instead (https://github.com/celestiaorg/iavl/issues/1). This is because due to the assumptions made by store V1 including range queries and versioning etc, making store V1 work well with SMT is more work than expected so we've decided to pursue IAVL for now.

**Comment by roysc** [comment]
> (if so can this PR be against the lazy branch rather than master)?

Sorry, I didn't see that branch - this branch is just somewhat cleaned up, I removed a caching MapStore which isn't quite relevant, renamed some minor things and added some comments. This is a rebase, so targeting the original head might be a mess, but I will check.

**Comment by musalbas** [comment]
@roysc does this PR remove value storage? I recall in the original lazy implementation in vulcanize, value storage was removed. But it looks like you added back in this PR, if I understand correctly (from a cursory look)? Would the code in the README still work?

File: hasher.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	"hash"
+)
+
+var (
+	leafPrefix  = []byte{0}
+	innerPrefix = []byte{1}
+	extPrefix   = []byte{2}
+)
+
+var _ PathHasher = (*pathHasher)(nil)
```
Comment (Olshansk) [inline hasher.go:14]:
```suggestion
var _ PathHasher = &pathHasher{}
```

Ditto below

File: options.go [hunk 2]
```diff
@@ -1,4 +1,12 @@
 package smt
 
-// Option is a function that configures SMT.
-type Option func(*SparseMerkleTree)
+// Option is a function that configures SparseMerkleTree.
+type Option func(*SMT)
+
+func SetPathHasher(ph PathHasher) Option {
```
Comment (Olshansk) [inline options.go]: NIT: Consider using `With` just to make it more idiomatic.

File: hasher.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	zeroValue []byte
+}
+type pathHasher struct {
+	treeHasher
+}
+type valueHasher struct {
+	treeHasher
+}
+
+func newTreeHasher(hasher hash.Hash) *treeHasher {
```
Comment (Olshansk) [inline hasher.go:42]: Publically exposing `NewTreeHasher` and also adding a `NewPathHash` would be helpful for upstream libraries using this.

**Review by Olshansk — COMMENTED** [review]
Did not do a deep dive into this PR but just left a few flyby NIT comments.

@roysc Do you plan on pushing this through? I could do a deep review or just submit a follow-up PR after this is merged if I see any areas of improvement.

@musalbas 

> does this PR remove value storage? 

Given the lack of `value` mapStore I believe the answer is yes, but again, I did not dive in.

Screen Shot 2022-11-09 at 1 09 30 PM

> Would the code in the README still work?

It does node. The value MapStore no longer needs to be passed in, the method signature change to `NewSMT`, and the options aren't captured. 

As someone who will use this after the PR is merged, I could add clarifying examples on the options available and such.

File: hasher.go (hunk 3 above)
Comment (roysc) [reply hasher.go:42]: the `treeHasher` interface is internal-only, and only the path/value hashers are exposed to keep the concerns separate. So I could expose constructors for those, but the interfaces are trivial anyway.

**Comment by roysc** [comment]
Sorry for the delay on this, it was sidelined by prioritization and time off. I'd like to wrap it up if possible, and to that end I've pushed an additional cleanup/refactor with a README update.

> @roysc does this PR remove value storage?

Yes it does, but we can easily add a type that wraps value storage with the tree. (Alternatively, the raw value can be stored directly in the leaf by using an identity function as the value hasher.)

What's the likelihood of this being reviewed in the near future? If no movement is likely for now (or you just plan to use your own branch), I will probably just close this PR.

**Comment by musalbas** [comment]
Thanks for this @roysc. Given we're not currently using SMT at Celestia, there's no one really currently maintaining this repo.

My suggestion is that given that your PR breaks the current interface, that you merge this into vulcanize/smt, and then perhaps I can add to the README of this repo that this repo isn't currently being maintained, and point them to vulcanize/smt as a more recently maintained version. Is vulcanize currently using and maintaining vulcanize/smt? Alternatively, I think it's fine to just leave this PR open in case we start maintaining this repo again, and for future users to see.

**Comment by roysc** [comment]
Thanks @musalbas. I think the SMT is basically shelved for vulcanize for now as well, but could possibly be experimented with in the future, so pointing to that fork makes sense. I'll close this PR though, just for tidiness purposes.

**Comment by Olshansk** [comment]
@musalbas @roysc  We're using this library at [Pocket](https://github.com/pokt-network/pocket) and plan on taking it to production as well. It probably won't be for a month or two, but what I was thinking was:

1. We can fork this library
2. Use this PR (update docs, test, etc...)
3. Optionally merged it back here to avoid maintaining a fork

**Comment by roysc** [comment]
@Olshansk That works, we will maintain our fork at Vulcanize and I can take up your PR once it's ready.

**Comment by musalbas** [comment]
We intend to deprecate/archive this repo as we don't have bandwidth to review PRs and maintain this repo. So @Olshansk if you plan to actively use and maintain SMT, I suggest maintaining your own fork for now.

**Comment by Olshansk** [comment]
> We intend to deprecate/archive this repo as we don't have bandwidth to review PRs and maintain this repo. So @Olshansk if you plan to actively use and maintain SMT, I suggest maintaining your own fork for now.

👍  I've created https://github.com/pokt-network/smt.

Will revisit potential collaboration opportunities when our fork is more mature & battle-tested.

**Pull request closed without merging**